adapter/stackdriver/log/log.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2017 the Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"text/template"
+	"time"
+
+	"cloud.google.com/go/logging"
+	xctx "golang.org/x/net/context"
+	"google.golang.org/api/option"
+
+	"istio.io/mixer/adapter/stackdriver/config"
+	"istio.io/mixer/adapter/stackdriver/helper"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/pkg/pool"
+	"istio.io/mixer/template/logentry"
+)
+
+type (
+	makeClientFn func(ctx xctx.Context, projectID string, opts ...option.ClientOption) (*logging.Client, error)
+
+	logFn func(logging.Entry)
+
+	builder struct {
+		makeClient makeClientFn
+		types      map[string]*logentry.Type
+	}
+
+	info struct {
+		labels []string
+		tmpl   *template.Template
+		req    *config.Params_LogInfo_HttpRequestMapping
+	}
+
+	handler struct {
+		now func() time.Time // used for testing
+
+		l      adapter.Logger
+		client io.Closer
+		log    logFn
+		info   map[string]info
+	}
+)
+
+var (
+	_ logentry.HandlerBuilder = &builder{}
+	_ logentry.Handler        = &handler{}
+)
+
+// NewBuilder returns a builder implementing the logentry.HandlerBuilder interface.
+func NewBuilder() logentry.HandlerBuilder {
+	return &builder{makeClient: logging.NewClient}
+}
+
+func (b *builder) ConfigureLogEntryHandler(types map[string]*logentry.Type) error {
+	b.types = types
+	return nil
+}
+
+func (b *builder) Build(c adapter.Config, env adapter.Env) (adapter.Handler, error) {
+	logger := env.Logger()
+	cfg := c.(*config.Params)
+	infos := make(map[string]info)
+	for name, log := range cfg.LogInfo {
+		_, found := b.types[name]
+		if !found {
+			logger.Infof("configured with log info about %s which is not an Istio log", name)
+			continue
+		}
+		tmpl, err := template.New(name).Parse(log.PayloadTemplate)
+		if err != nil {
+			_ = logger.Errorf("failed to evaluate template for log %s, skipping: %v", name, err)
+			continue
+		}
+		infos[name] = info{
+			labels: log.LabelNames,
+			tmpl:   tmpl,
+			req:    log.HttpMapping,
+		}
+	}
+
+	client, err := b.makeClient(context.Background(), cfg.ProjectId, helper.ToOpts(cfg)...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stackdriver logging client: %v", err)
+	}
+	return &handler{log: client.Logger("istio-mixer").Log, client: client, now: time.Now, l: logger}, nil
```

Comment:
should the `logName` be configurable?  Or are we going to enforce always logging to `istio-mixer` ?

Reply:
My understanding is that this winds up being the log ID, which makes up the resource name. It identifies all the entries, so with this name all Stackdriver logs produced by Mixer belong to the same log collection. We should probably restructure this to create and hold on to a logger per Istio log name. We might even want to prefix the log name with "istio-" or similar.